trio/_core/tests/test_watchdog.py

```diff
@@ -0,0 +1,30 @@
+import threading
+import time
+from io import StringIO
+
+import contextlib
+
+from ..tests.tutil import slow
+from .. import _run
+
+
+@slow
+async def test_watchdog():
+    watchdog = _run.GLOBAL_RUN_CONTEXT.watchdog
+    target = StringIO()
+    with contextlib.redirect_stderr(target):
+        time.sleep(7)
+
+    assert target.getvalue().startswith(
+        "Trio Watchdog has not received any "
+        "notifications in 5 seconds, main "
+        "thread is blocked!"
+    )
+
+    # checkpoint to ensure the watchdog gets a notification, sees that its
+    # dead, and winds down its thread
+    watchdog.stop()
+    await _run.checkpoint()
+    time.sleep(6)  # ensure if the watchdog is waiting for 5s, it wakes
+    await _run.checkpoint()
+    assert not watchdog._thread.is_alive()
```

Comment:
This looks very delicate... calling `self._event.set()` from `Watchdog.stop()` will help, but even so...

Here's another idea: make the watchdog thread *not* a daemon, and then join it in `Watchdog.stop`. That way every time we call `trio.run`, we're implicitly checking the watchdog shutdown logic (and also making sure we don't leave stray background threads running, even temporarily). What do you think?